engine/table/src/main/java/io/deephaven/engine/table/impl/sources/ring/AbstractRingChunkSource.java

```diff
@@ -0,0 +1,389 @@
+        } else {
+            final long skipRows = logicalFillSize - capacity;
+            physicalRows = srcKeys.subSetByPositionRange(skipRows, logicalFillSize);
+            physicalStartRingIx = nextRingIx + skipRows;
+            physicalFillSize = capacity;
+        }
+        try {
+            // [0, capacity)
+            final int fillIndex1 = keyToRingIndex(physicalStartRingIx);
+            // (0, capacity]
+            final int fillMax1 = capacity - fillIndex1;
+            // fillSize1 + fillSize2 = physicalFillSize
+            final int fillSize1 = Math.min(fillMax1, physicalFillSize);
+            final int fillSize2 = physicalFillSize - fillSize1;
+            if (fillSize2 == 0) {
+                src.fillChunk(fillContext, ring(fillIndex1, fillSize1), physicalRows);
+            } else {
+                // might be nice if there was a "split"
+                // (could be more efficient than calling subSetByPositionRange twice)
+                try (final RowSet rows1 = physicalRows.subSetByPositionRange(0, fillSize1)) {
+                    src.fillChunk(fillContext, ring(fillIndex1, fillSize1), rows1);
+                }
+                try (final RowSet rows2 = physicalRows.subSetByPositionRange(fillSize1, fillSize1 + fillSize2)) {
+                    src.fillChunk(fillContext, ring(0, fillSize2), rows2);
+                }
+            }
+        } finally {
+            if (hasSkippedRows) {
+                physicalRows.close();
+            }
+        }
+        nextRingIx += logicalFillSize;
+    }
+
+    @Override
+    public final Chunk<Values> getChunk(@NotNull GetContext context, @NotNull RowSequence rowSequence) {
+        if (rowSequence.isEmpty()) {
+            return getChunkType().getEmptyChunk();
+        }
+        if (rowSequence.isContiguous()) {
+            return getChunk(context, rowSequence.firstRowKey(), rowSequence.lastRowKey());
+        }
+        final WritableChunk<Values> chunk = DefaultGetContext.getWritableChunk(context);
+        fillChunk(DefaultGetContext.getFillContext(context), chunk, rowSequence);
+        return chunk;
+    }
+
+    @Override
+    public final Chunk<Values> getChunk(@NotNull GetContext context, long firstKey, long lastKey) {
+        // This check should not be necessary given precondition on getChunk
+        if (!containsRange(firstKey, lastKey)) {
+            throw new IllegalStateException(
+                    String.format("getChunk precondition broken, invalid range. requested=[%d, %d], available=[%d, %d]",
+                            firstKey, lastKey, firstKey(), lastKey()));
+        }
+        final int firstRingIx = keyToRingIndex(firstKey);
+        final int secondRingIx = keyToRingIndex(lastKey);
+        if (firstRingIx <= secondRingIx) {
+            // Optimization when we can return a contiguous view
+            return ring(firstRingIx, secondRingIx - firstRingIx + 1); // todo: is it bad that this is writable?
```

Comment:
Regarding your to-do, it's OK. Callers are *not* allowed to modify the result of a `getChunk`.